relations/jf/src/note.rs

```diff
@@ -102,16 +107,19 @@ impl NoteGadget for PlonkCircuit<CircuitField> {
 impl PublicInput for SourcedNote {
     fn public_input(&self) -> Vec<CircuitField> {
         match self.note_type {
-            NoteType::Spend => {
-                vec![convert_hash(self.nullifier)]
-            }
             NoteType::Deposit => {
                 vec![
-                    convert_hash(self.note),
+                    convert_array(self.note),
                     self.token_id.into(),
                     self.token_amount.into(),
                 ]
             }
+            NoteType::Spend => {
+                vec![convert_array(self.nullifier)]
+            }
+            NoteType::Redeposit => {
+                vec![convert_array(self.note), self.token_id.into()]
+            }
```

Comment:
Alternative design would be to split `SourceNote` into 3 separate structs: `Deposit`, `Spend`, `Redeposit` with an associated type `PublicInput` and get the typesafety wrt to different sets of circuit field elements.

Reply:
Here, the return type is opaque to the caller who has no idea what it us, why it is etc. If there was a single type representing the public inputs for that specific relation then there's no confusion - i.e. you can't supply `WithdrawRelation#PublicInput` where `DepositRelation#PublicInput` is expected while you can do that with `Vec<CircuitField>` 🤷

Reply:
In general this is obviously true but in the way we use public input this doesn't seem to give much, right?

Reply:
I just think we should write "better" code always, not only when it makes sense.

Doesn't have to be addressed in this PR obvs.